### PR TITLE
Optimization for moving least squares

### PR DESCRIPTION
**Pull request by mrlag31**
This PR modifies the SVD pseudo-inverse and the coefficients computation, its main goal was to reduce the amount of global memory used, use scratch memory and compound the kernels.

### CODE REVIEW

**Comment by aprokop** [comment]
@mrlag31 Can you please explain in more details what the difference is?

**Comment by mrlag31** [comment]
- The previous SVD implementation used two extra square matrices (one for the diagonalisation, one for the unit matrices). Now, it uses a vector and a square matrix, the diagonalisation is done on the input matrix.
- The coefficients computation is now batched per-thread using a team policy. All of the previous kernels were compounded into a single one. All of the temporary matrices are now placed in shared memory.

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 1]
```diff
@@ -26,13 +26,206 @@
...
+    val += LocalSVDDiag::shmem_size(poly_size);
+    val += LocalSVDUnit::shmem_size(poly_size, poly_size);
+    return team_size * val;
+  }
+
+  template <typename TeamMember>
+  KOKKOS_FUNCTION void operator()(TeamMember member) const
+  {
+    int const rank = member.team_rank();
+    int const size = member.team_size();
```
Comment (dalg24) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:84]: Do you really want these local variable?  They are only used to determine the target point index.

Comment (dalg24) [reply src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:84]: Or for readability defer to a function that takes the index as argument.

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 2]
```diff
@@ -26,13 +26,206 @@
...
+
+  using CoefficientsType = typename Coefficients::non_const_value_type;
+
+  static constexpr int dimension = GeometryTraits::dimension_v<SourcePoint>;
+  static constexpr int degree = PolynomialDegree::value;
+  static constexpr int poly_size = polynomialBasisSize<dimension, degree>();
+
+  template <typename T>
+  using UnmanagedView =
+      Kokkos::View<T, ScratchMemorySpace, Kokkos::MemoryUnmanaged>;
```
Comment (dalg24) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:47]: The relevant part in the name is "scratch" not "unmanaged".

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 3]
```diff
@@ -26,13 +26,206 @@
...
+
+  using LocalSourcePoints = Kokkos::Subview<SourcePoints, int, Kokkos::ALL_t>;
+  using LocalPhi = UnmanagedView<CoefficientsType *>;
+  using LocalVandermonde = UnmanagedView<CoefficientsType **>;
+  using LocalMoment = UnmanagedView<CoefficientsType **>;
+  using LocalSVDDiag = UnmanagedView<CoefficientsType *>;
+  using LocalSVDUnit = UnmanagedView<CoefficientsType **>;
+  using LocalCoefficients = Kokkos::Subview<Coefficients, int, Kokkos::ALL_t>;
```
Comment (dalg24) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:55]: Why is it relevant to say "local" here?

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 4]
```diff
@@ -26,13 +26,206 @@
...
+  std::size_t team_shmem_size(int const team_size) const
+  {
+    std::size_t val = 0;
+    val += LocalPhi::shmem_size(_num_neighbors);
+    val += LocalVandermonde::shmem_size(_num_neighbors, poly_size);
+    val += LocalMoment::shmem_size(poly_size, poly_size);
+    val += LocalSVDDiag::shmem_size(poly_size);
+    val += LocalSVDUnit::shmem_size(poly_size, poly_size);
+    return team_size * val;
+  }
```
Comment (dalg24) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:78]: Prefer the `TeamPolicy::set_scratch_size` method

**Review by dalg24 — COMMENTED** [review]
Drive-by comments

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 5]
```diff
@@ -26,13 +26,206 @@
...
+    // Where:
+    // - p(x) is the polynomial basis of point x (line vector).
+    // - P is the multidimensional Vandermonde matrix built from the source
+    //   points, i.e., each line is the polynomial basis of a source point.
+    // - PHI is the diagonal weight matrix / CRBF evaluated at each source
+    // point.
+
+    // We first change the origin of the evaluation to be at the target point.
+    // This lets us use p(0) which is [1 0 ... 0].
+    sourceRecenteringKernel(target_point, source_points);
```
Comment (masterleinad) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:111]: For me, a "kernel" is whatever is launched from the host, i.e., a `__global__` function in `Cuda`-speech.
```suggestion
    sourceRecentering(target_point, source_points);
```

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 6]
```diff
@@ -26,13 +26,206 @@
...
+    // Now, the moment has [P^T.PHI.P]^-1
+
+    // Finally, the result is produced by computing p(0).[P^T.PHI.P]^-1.P^T.PHI
+    coefficientsKernel(phi, vandermonde, moment, coefficients);
+  }
+
+  Kokkos::TeamPolicy<ExecutionSpace>
+  make_policy(ExecutionSpace const &space) const
+  {
+    Kokkos::TeamPolicy<ExecutionSpace> policy(space, 1, Kokkos::AUTO);
```
Comment (masterleinad) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:135]:
```suggestion
    Kokkos::TeamPolicy<ExecutionSpace> dummy_policy(space, 1, Kokkos::AUTO);
```

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 7]
```diff
@@ -26,13 +26,206 @@
...
+    int team_rec =
+        policy.team_size_recommended(*this, Kokkos::ParallelForTag{});
+    int mod = _num_targets % team_rec;
+    int div = (_num_targets - mod) / team_rec;
+    int league_rec = div + ((mod == 0) ? 0 : 1);
+    return Kokkos::TeamPolicy<ExecutionSpace>(space, league_rec, team_rec);
```
Comment (masterleinad) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:141]:
```suggestion
    int team_size =
        policy.team_size_recommended(*this, Kokkos::ParallelForTag{});
    int league_size = (_num_targets + team_size-1)/team_size;
    return Kokkos::TeamPolicy<ExecutionSpace>(space, league_size, team_size);
```

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 8]
```diff
@@ -69,167 +264,24 @@ movingLeastSquaresCoefficients(ExecutionSpace const &space,
...
-      });
-
-  Kokkos::Profiling::popRegion();
-  return coeffs;
+      source_points.extent_int(0), source_points.extent_int(1));
+
+  MovingLeastSquaresCoefficientsKernel<CRBFunc, PolynomialDegree, SourcePoints,
+                                       TargetAccess, decltype(coefficients),
+                                       ExecutionSpace>
+      kernel(space, target_access, source_points, coefficients);
```
Comment (masterleinad) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:279]: Can we reorder the template arguments so we don't need to specify all of them here?

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp (hunk 1 above)
Comment (mrlag31) [reply src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:84]: They were used elsewhere before, I will remove them.

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp (hunk 3 above)
Comment (mrlag31) [reply src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:55]: It is to signify "local to the current target"

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 9]
```diff
@@ -69,167 +264,24 @@ movingLeastSquaresCoefficients(ExecutionSpace const &space,
   static_assert(dimension == GeometryTraits::dimension_v<TargetPoint>,
                 "target and source points must have the same dimension");
 
-  int const num_targets = target_access.size();
-  int const num_neighbors = source_points.extent(1);
+  // The number of source groups must be correct
+  KOKKOS_ASSERT(target_access.size() == source_points.extent_int(0));
```
[the remark this answers is not in the document]
Comment (mrlag31) [reply src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:268]: I will wrap `target_access.size` around an unsigned integer and use the regular `extent` call

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp (hunk 8 above)
Comment (mrlag31) [reply src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:279]: I'll put the CRBF and PolynomialDegree at the end and set defaults. The 4 others must be determined by the outside.

File: src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp [hunk 10]
```diff
@@ -26,13 +26,204 @@
...
+  template <typename T>
+  using ScratchView =
+      Kokkos::View<T, ScratchMemorySpace, Kokkos::MemoryUnmanaged>;
+
+  using LocalSourcePoints = Kokkos::Subview<SourcePoints, int, Kokkos::ALL_t>;
+  using LocalPhi = ScratchView<CoefficientsType *>;
+  using LocalVandermonde = ScratchView<CoefficientsType **>;
+  using LocalMoment = ScratchView<CoefficientsType **>;
+  using LocalSVDDiag = ScratchView<CoefficientsType *>;
+  using LocalSVDUnit = ScratchView<CoefficientsType **>;
```
Comment (dalg24) [inline src/interpolation/details/ArborX_InterpDetailsMovingLeastSquaresCoefficients.hpp:55]:
```suggestion
  using LocalVandermonde = ScratchView<CoefficientsType *[poly_size]>;
  using LocalMoment = ScratchView<CoefficientsType[poly_size][poly_size]>;
  using LocalSVDDiag = ScratchView<CoefficientsType [poly_size]>;
  using LocalSVDUnit = ScratchView<CoefficientsType[poly_size][poly_size]>;
```

**Review by aprokop — COMMENTED** [review]
I'm not sure about renaming in `symmetricPseudoInverseSVDKernel`. I don't think the new names make it any clearer, and would prefer the old names.

**Comment by mrlag31** [comment]
> I'm not sure about renaming in `symmetricPseudoInverseSVDKernel`. I don't think the new names make it any clearer, and would prefer the old names.

I felt like the old ones were harder to understand. The new ones directly explains, by their name, what they are in regards to the formula, whereas the old ones were a bit too abstract.

**Review by aprokop — APPROVED** [review]
I'm ok with the current state.

**Pull request merged**